src/transform/src/canonicalize_mfp.rs

```diff
@@ -75,14 +76,49 @@ impl CanonicalizeMfp {
                 // Maximize number of predicates that are sped using a single index.
                 .max_by_key(|(key, _val)| key.len());
             if let Some((key, val)) = key_val {
-                *relation = MirRelationExpr::Join {
-                    implementation: mz_expr::JoinImplementation::PredicateIndex(
-                        *id,
+                // We transform the Get into a join with a constant collection.
+                // (For now, the constant collection always has only 1 element.)
+                // E.g.: we go from something like
+                // `SELECT f1, f2, f3 FROM t WHERE t.f1 = lit1 AND t.f2 = lit2
+                // to
+                // `SELECT f1, f2, f3 FROM t, (SELECT * FROM (VALUES (lit1, lit2))) as filter_list
+                //  WHERE t.f1 = filter_list.column1 AND t.f2 = filter_list.column2`
+
+                let inp_id = id.clone();
+                let inp_typ = relation.typ();
+                let filter_list = MirRelationExpr::Constant {
+                    rows: Ok(vec![(val.clone(), 1)]),
+                    typ: mz_repr::RelationType {
+                        column_types: key.iter().map(|e| e.typ(&inp_typ)).collect(),
+                        keys: vec![],
+                    },
+                };
+
+                let all_cols_filter_list =
+                    MirScalarExpr::columns(&(0..key.len()).collect::<Vec<usize>>()[..]);
+
+                let join = MirRelationExpr::Join {
+                    inputs: vec![relation.clone().arrange_by(&[key.clone()]), filter_list],
+                    equivalences: key
+                        .iter()
+                        .enumerate()
+                        .map(|(i, e)| {
+                            vec![(*e).clone(), MirScalarExpr::column(i + inp_typ.arity())]
+                        })
+                        .collect(),
+                    // The filter_list has to be the starting collection, at least until #14059 is fixed
+                    implementation: PredicateIndex(
+                        inp_id,
                         key.clone(),
                         val,
+                        (1, Some(all_cols_filter_list)),
+                        vec![(0, key)],
```

Comment:
I don't get the point of adding these two columns, can you explain?

Reply:
The last 2 fields of `PredicateIndex` are the 2 fields of `Differential`. I have these fields so that it's easy in the lowering to use the same code path as `Differential`. Btw. I might reorganize the fields of `PredicateIndex` a bit in the next PR. I just didn't want to touch the fast path in this PR, but the next one will do some stuff there anyway. Currently, only the first 3 fields are used in the fast path, and only the last 2 fields are used in the lowering.

Reply:
The value that is set in the final field (`vec![(0, key)]`) can be derived from the structure, so you can do this at the client side whenever you need it in this form rather than duplicating it in the definition.

Reply:
The [client code in the lowering](https://github.com/ggevay/materialize/blob/4d2b84514c08dd8fdc479cc7d92f4fa5ac6f3682/src/compute-client/src/plan/mod.rs#L1094) would become much more hairy. Currently it's just a `|` pattern between `PredicateIndex` and `Differential`. Both sides of the `|` have to bind `order`, and I don't know of a good way to derive `vec![(0, key)]`  inside the pattern from `key`. So I would say that having this minor duplication in the definition is the smaller bad.

Reply:
Or maybe do you know some Rust pattern matching magic to solve this?

Reply:
You are trading off precision in `PredicateIndex` for the sake of ergonomics of one of it's consumers.

Prior to that change, every instance of `PredicateIndex` was valid. After the change, every producer will need to worry about adhering to the new constraint between the `key` field and the newly added fields. This will be OK if we were consistently relying on [smart constructors](https://wiki.haskell.org/Smart_constructors) for LIR expressions in our codebase, but we don't.

I wouldn't want to debug problems caused by mis-constructed `PredicateIndex` instances. Rather, I would just split the two patterns and duplicate this code and in the case of `PredicateIndex()` match compute the `start` and `order`.

```rust
// Extract temporal predicates as joins cannot currently absorb them.
let (plan, missing) = match implementation {
    PredicateIndex(_, key, _)  => {
        let source_arrangement = input_keys[*start].arbitrary_arrangement();
        let (ljp, missing) = LinearJoinPlan::create_from(
            1,
            source_arrangement,
            equivalences,
            &vec![(0, key.clone())],
            input_mapper,
            &mut mfp,
            &input_keys,
        );
        (JoinPlan::Linear(ljp), missing)
    }
```

Reply:
Ok, changed it.